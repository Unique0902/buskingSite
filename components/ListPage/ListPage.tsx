import React, { useState } from 'react';
import { calculateTotalPageNum } from '../../utils/calculate';
import PagingBar from '../Table/PagingBar';

interface Props<T> {
  resultTotalNum: number;
  resultNumPerPage: number;
  pageDataArr: T[];
  renderData: (data: T, index: number) => React.JSX.Element;
  // handleChangePage: (pageNum: number) => void;
}

const ListPage = <T,>({
  resultTotalNum,
  resultNumPerPage,
  pageDataArr,
  renderData,
}: // handleChangePage,
Props<T>) => {
  const [nowPageNum, setNowPageNum] = useState<number>(1);
  const handlePlusPageNum = () => {};
  const handleMinusPageNum = () => {};
  return (
    <>
      {pageDataArr.map((data, idx) => renderData(data, idx))}
      <PagingBar
        totalPageNum={calculateTotalPageNum(resultTotalNum, resultNumPerPage)}
        pageNum={nowPageNum}
        onPagePlus={handlePlusPageNum}
        onPageMinus={handleMinusPageNum}
      />
    </>
  );
};

export default ListPage;
