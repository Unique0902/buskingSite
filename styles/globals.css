@tailwind base;
@tailwind components;
@tailwind utilities;
/* TODO: 글꼴 불러오는 로딩 줄이거나 없애기 */
* {
  box-sizing: border-box;
  font-family: 'Pretendard';
}

@font-face {
  font-family: 'Pretendard';
  font-weight: 45 920;
  font-style: normal;
  font-display: swap;
  src: url('../assets/fonts/PretendardVariable.woff2')
    format('woff2-variations');
}

li {
  list-style: none;
}
body {
  margin: 0;
  @apply text-black dark:text-white dark:bg-slate-600;
}
p,
h1,
h2 {
  margin: 0;
}
ul {
  padding: 0;
}
