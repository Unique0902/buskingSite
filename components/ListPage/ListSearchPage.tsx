import React, { useState } from 'react';
import ListPage from './ListPage';

interface Props<T> {
  pageDataInform: {
    resultTotalNum: number;
    resultNumPerPage: number;
  };
  pageDataArr: T[];
  renderData: (data: T, index: number) => React.JSX.Element;
  handleChangePage: (pageNum: number) => void;
  renderNoData: () => React.JSX.Element;
}

const ListSearchPage = <T,>({
  pageDataInform,
  pageDataArr,
  renderData,
  handleChangePage,
  renderNoData,
}: Props<T>) => {
  const [searchWord, setSearchWord] = useState({ name: '', category: '' });
  const [nowPageNum, setNowPageNum] = useState<number>(1);
  return (
    <div>
      <ListPage
        pageDataInform={pageDataInform}
        pageDataArr={pageDataArr}
        renderData={renderData}
        handleChangePage={handleChangePage}
        renderNoData={renderNoData}
      />
    </div>
  );
};

export default ListSearchPage;
