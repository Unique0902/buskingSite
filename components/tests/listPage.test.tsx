import { describe, expect, it, jest } from '@jest/globals';
import renderer from 'react-test-renderer';
import userEvent from '@testing-library/user-event';
import ListPage from '../ListPage/ListPage';
import { render, screen } from '@testing-library/react';

describe('List Page Component Test', () => {
  type TestData = {
    title: string;
    description: string;
  };

  const testPageDataArr: TestData[] = [
    { title: 'title1', description: 'description1' },
    { title: 'title2', description: 'description2' },
  ];

  const testPageDataArr2: TestData[] = [
    { title: 'title1', description: 'description1' },
    { title: 'title2', description: 'description2' },
  ];

  const bigTestPageDataArr: TestData[] = [
    { title: 'title1', description: 'description1' },
    { title: 'title2', description: 'description2' },
    { title: 'title2', description: 'description2' },
    { title: 'title2', description: 'description2' },
    { title: 'title2', description: 'description2' },
    { title: 'title2', description: 'description2' },
    { title: 'title2', description: 'description2' },
    { title: 'title2', description: 'description2' },
    { title: 'title2', description: 'description2' },
    { title: 'title2', description: 'description2' },
  ];

  const bigTestPageDataArr2: TestData[] = [
    { title: 'title1', description: 'description1' },
    { title: 'title2', description: 'description2' },
    { title: 'title2', description: 'description2' },
    { title: 'title2', description: 'description2' },
    { title: 'title2', description: 'description2' },
    { title: 'title2', description: 'description2' },
    { title: 'title2', description: 'description2' },
    { title: 'title2', description: 'description2' },
    { title: 'title2', description: 'description2' },
    { title: 'title2', description: 'description2' },
  ];

  it('test rendering correctly', () => {
    const component = renderer.create(
      <ListPage<TestData>
        pageDataArr={testPageDataArr}
        renderData={(data, idx) => (
          <div key={data.title + idx}>{data.title}</div>
        )}
        pageDataInform={{
          resultTotalNum: testPageDataArr.length,
          resultNumPerPage: 6,
          totalDataArr: testPageDataArr,
        }}
        handleChangePage={() => {}}
      />
    );
    expect(component.toJSON()).toMatchSnapshot();
  });

  it('show given dataArr by given figure', () => {
    render(
      <ListPage<TestData>
        pageDataArr={testPageDataArr}
        renderData={(data, idx) => (
          <div key={data.title + idx}>{data.title}</div>
        )}
        pageDataInform={{
          resultTotalNum: testPageDataArr.length,
          resultNumPerPage: 6,
          totalDataArr: testPageDataArr,
        }}
        handleChangePage={() => {}}
      />
    );
    expect(screen.queryByText('title1')).not.toBeNull();
    expect(screen.queryByText('title2')).not.toBeNull();
  });

  it('when click pagingBarBtn, excute handleChangePage', async () => {
    const handleChangePage = jest.fn();
    render(
      <ListPage<TestData>
        pageDataArr={testPageDataArr}
        renderData={(data, idx) => (
          <div key={data.title + idx}>{data.title}</div>
        )}
        pageDataInform={{
          resultTotalNum: bigTestPageDataArr.length,
          resultNumPerPage: 2,
          totalDataArr: bigTestPageDataArr,
        }}
        handleChangePage={handleChangePage}
      />
    );
    const [minusBtn, plusBtn] = screen.queryAllByRole('button');
    await userEvent.click(plusBtn);
    expect(handleChangePage).toHaveBeenCalledWith(2);
    await userEvent.click(minusBtn);
    expect(handleChangePage).toHaveBeenCalledWith(1);
  });

  it('when pageDataArr length is bigger than resultNumPerPage', async () => {
    expect(() =>
      render(
        <ListPage<TestData>
          pageDataArr={testPageDataArr}
          renderData={(data, idx) => (
            <div key={data.title + idx}>{data.title}</div>
          )}
          pageDataInform={{
            resultTotalNum: bigTestPageDataArr.length,
            resultNumPerPage: 1,
            totalDataArr: bigTestPageDataArr,
          }}
          handleChangePage={() => {}}
        />
      )
    ).toThrow(
      'pageDataArr length must be same or smaller than resultNumPerPage'
    );
  });

  it('when no data', () => {
    render(
      <ListPage<TestData>
        pageDataArr={[]}
        renderData={(data, idx) => (
          <div key={data.title + idx}>{data.title}</div>
        )}
        pageDataInform={{
          resultTotalNum: 0,
          resultNumPerPage: 6,
          totalDataArr: [],
        }}
        handleChangePage={() => {}}
        renderNoData={() => <div>no data</div>}
      />
    );
    expect(screen.queryByText('no data')).not.toBeNull();
  });
  // TODO: totalDataArr 변화했을때 pageNum 0 된거 테스트하기

  it('when totalDataArr change, pageNum clear', async () => {
    const { rerender } = render(
      <ListPage<TestData>
        pageDataArr={testPageDataArr}
        renderData={(data, idx) => (
          <div key={data.title + idx}>{data.title}</div>
        )}
        pageDataInform={{
          resultTotalNum: bigTestPageDataArr.length,
          resultNumPerPage: 2,
          totalDataArr: bigTestPageDataArr,
        }}
        handleChangePage={() => {}}
      />
    );
    const [minusBtn, plusBtn] = screen.queryAllByRole('button');
    await userEvent.click(plusBtn);
    expect(screen.queryByText('2')).not.toBeNull();
    rerender(
      <ListPage<TestData>
        pageDataArr={testPageDataArr2}
        renderData={(data, idx) => (
          <div key={data.title + idx}>{data.title}</div>
        )}
        pageDataInform={{
          resultTotalNum: bigTestPageDataArr2.length,
          resultNumPerPage: 2,
          totalDataArr: bigTestPageDataArr2,
        }}
        handleChangePage={() => {}}
      />
    );

    expect(screen.queryByText('1')).not.toBeNull();
  });
});
