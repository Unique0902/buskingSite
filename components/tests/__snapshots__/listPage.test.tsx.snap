// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`List Page Component Test test rendering correctly 1`] = `
<ul
  className="p-1 bg-gray-800 rounded-xl"
>
  <div>
    title1
  </div>
  <div>
    title2
  </div>
  <div
    className="flex flex-row items-center justify-center mt-3"
  >
    <button
      className=" rounded-full hover:scale-110 px-4 py-2"
      onClick={[Function]}
    >
      <svg
        fill="white"
        height={18}
        viewBox="0 0 306 306"
        width={18}
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M247.35 35.7 211.65 0l-153 153 153 153 35.7-35.7L130.05 153z"
          fillRule="nonzero"
        />
      </svg>
    </button>
    <div
      className="flex flex-row items-center gap-2 text-xl font-normal text-white"
    >
      <p>
        1
      </p>
      <p>
        /
      </p>
      <p>
        1
      </p>
    </div>
    <button
      className=" rounded-full hover:scale-110 px-4 py-2"
      onClick={[Function]}
    >
      <svg
        fill="white"
        height={18}
        viewBox="0 0 306 306"
        width={18}
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M58.65 267.75 175.95 153 58.65 35.7 94.35 0l153 153-153 153z"
          fillRule="nonzero"
        />
      </svg>
    </button>
  </div>
</ul>
`;
